lib/babe/babe.go

```diff
@@ -336,158 +303,105 @@ func (b *Service) initiate() {
 		return
 	}
 
-	err := b.invokeBlockAuthoring()
-	if err != nil {
-		logger.Criticalf("block authoring error: %s", err)
+	// we should consider better error handling for this - we should
+	// retry to run the engine at some point (maybe the next epoch) if
+	// there's an error.
+	if err := b.runEngine(); err != nil {
+		logger.Criticalf("failed to run block production engine: %s", err)
 	}
 }
 
-func (b *Service) invokeBlockAuthoring() error {
-	epoch, err := b.epochState.GetCurrentEpoch()
+func (b *Service) initiateAndGetEpochHandler(epoch uint64) (*epochHandler, error) {
+	epochData, err := b.initiateEpoch(epoch)
 	if err != nil {
-		logger.Errorf("failed to get current epoch: %s", err)
-		return err
+		return nil, fmt.Errorf("failed to initiate epoch %d: %s", epoch, err)
 	}
 
-	for {
-		err := b.initiateEpoch(epoch)
-		if err != nil {
-			logger.Errorf("failed to initiate epoch %d: %s", epoch, err)
-			return err
-		}
-
-		logger.Debugf("initiated epoch with threshold %s, randomness 0x%x and authorities %v",
-			b.epochData.threshold, b.epochData.randomness[:], b.epochData.authorities)
+	logger.Debugf("initiated epoch with threshold %s, randomness 0x%x and authorities %v",
+		epochData.threshold, epochData.randomness[:], epochData.authorities)
 
-		epochStartSlot, err := b.waitForEpochStart(epoch)
-		if err != nil {
-			logger.Errorf("failed to wait for epoch %d start: %s", epoch, err)
-			return err
-		}
-
-		// calculate current slot
-		startSlot := getCurrentSlot(b.slotDuration)
-		intoEpoch := startSlot - epochStartSlot
-
-		// if the calculated amount of slots "into the epoch" is greater than the epoch length,
-		// we've been offline for more than an epoch, and need to sync. pause BABE for now, syncer will
-		// resume it when ready
-		if b.epochLength <= intoEpoch && !b.dev {
-			logger.Debugf(
-				"pausing BABE, need to sync since we have %d slots (start slot %d) into the epoch starting at %d",
-				intoEpoch, startSlot, epochStartSlot)
-			return b.Pause()
-		}
+	epochStartSlot, err := b.epochState.GetStartSlotForEpoch(epoch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start slot for current epoch %d: %s", epoch, err)
+	}
 
-		if b.dev {
-			intoEpoch = intoEpoch % b.epochLength
-		}
+	return newEpochHandler(epoch,
+		epochStartSlot,
+		epochData,
+		b.constants,
+		b.handleSlot,
+		b.keypair,
+	)
+}
 
-		logger.Infof("current epoch %d has %d slots", epoch, intoEpoch)
+func (b *Service) runEngine() error {
+	epoch, err := b.epochState.GetCurrentEpoch()
+	if err != nil {
+		return fmt.Errorf("failed to get current epoch: %s", err)
+	}
 
-		// get start slot for current epoch
-		nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-		if err != nil {
-			logger.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			return err
-		}
+	for {
+		err = func() error {
```

Comment:
can we have this one as a separate global scope function?

Reply:
done!